src/ast/FunctionalConstraint.h

```diff
@@ -0,0 +1,128 @@
+/*
+ * Souffle - A Datalog Compiler
+ * Copyright (c) 2020 The Souffle Developers. All rights reserved
+ * Licensed under the Universal Permissive License v 1.0 as shown at:
+ * - https://opensource.org/licenses/UPL
+ * - <souffle root>/licenses/SOUFFLE-UPL.txt
+ */
+
+/************************************************************************
+ *
+ * @file FunctionalConstraint.h
+ *
+ * Defines the functional constraint class
+ *
+ ***********************************************************************/
+
+#pragma once
+
+#include "ast/Argument.h"
+#include "ast/Constraint.h"
+#include "ast/Node.h"
+#include "ast/Variable.h"
+#include "ast/utility/NodeMapper.h"
+#include "parser/SrcLocation.h"
+#include "souffle/utility/ContainerUtil.h"
+#include "souffle/utility/MiscUtil.h"
+#include "souffle/utility/StreamUtil.h"
+#include <cassert>
+#include <iostream>
+#include <memory>
+#include <string>
+#include <utility>
+#include <vector>
+
+namespace souffle::ast {
+
+/**
+ * @class FunctionalConstraint
+ * @brief Functional constraint (choice construct) class
+ *
+ * Representing a functional dependency (choice construct)
+ * For relation (x, y), we have x -> y, a function from x to y, i.e. x uniquely identifies some y value
+ */
+class FunctionalConstraint : public Constraint {
+public:
+    FunctionalConstraint(VecOwn<Variable> keys, SrcLocation loc = {})
+            : Constraint(std::move(loc)), keys(std::move(keys)) {}
+
+    FunctionalConstraint(Own<Variable> key, SrcLocation loc = {}) : Constraint(std::move(loc)) {
+        keys.push_back(std::move(key));
+    }
+
+    /** get keys */
+    std::vector<Variable*> getKeys() const {
+        return toPtrVector(keys);
+    }
+
+    /** get arity of the keys (i.e. number of source nodes: (x,y)->z has an arity of 2) */
+    size_t getArity() const {
+        return keys.size();
+    }
+
+    std::vector<const Node*> getChildNodes() const override {
+        std::vector<const Node*> res;
+        for (auto& cur : keys) {
+            res.push_back(cur.get());
+        }
+        /* res.push_back(rhs.get()); */
+        return res;
+    }
+
+    FunctionalConstraint* clone() const override {
+        VecOwn<Variable> newKeys;
+        for (const auto& key : keys) {
+            newKeys.push_back(Own<Variable>(key->clone()));
+        }
+        auto* res = new FunctionalConstraint(std::move(newKeys));
+        res->setSrcLoc(getSrcLoc());
+        return res;
+    }
+
+public:
+    void print(std::ostream& os) const override {
```

Comment:
Can you ensure that the printed AST is identical to the parsed AST, e.g., `keys a,b,(c,d)`?